plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Debug/Debugger.hs

```diff
@@ -0,0 +1,267 @@
+import Control.Monad.State
+import Control.Monad.Trans.Free
+import Control.Lens hiding (Context)
+import Prettyprinter
+import Data.Sequence as Seq
+import Data.Coerce
+import GHC.Exts
+
+-- usually it is a line, but let's treat more generally as a srcspan
+newtype Breakpoint = Breakpoint { unBreakpoint :: SrcSpan }
+    deriving newtype (Show, Pretty)
+
+-- | The terms that we can debug
+type DTerm uni fun = Term NamedDeBruijn uni fun SrcSpans
+
+-- commands that the debugger can receive fromthe debug-client (tui,cli,test,etc)
+data Cmd
+  =
+  -- MAYBE: have a Load DTerm command; would require an initial invalid debugstate, i.e.  Maybe DebugState
+  Reload -- ^ reset back to start of initial term
+  -- small and big steps
+  | Step -- ^ step the cek once
+  | Next -- ^ step over aka next
+  | Finish -- ^ run to the end of current function
+  | ContinueBreak -- ^ continue to the end or until breakpoint reached
+  | ContinueIgnore -- ^ continue to the end by temporarily ignoring breakpoints
+  -- breakpoints
+  | BreakAdd Breakpoint -- ^ create a new breakpoint
+  | BreakDel Word -- ^ delete a breakpoint by index no.
+  | BreakDelAll -- ^ convenience function
+  | BreakList -- ^ list breakpoints
+  -- variables
+  | Print NamedDeBruijn
+  | PrintAll -- ^ all debruijn in the env
+  | Discharge NamedDeBruijn -- ^ dischargeCekValue
+  | DischargeAll -- ^ forall v in env . dischargeCekValue
+  | DischargeControl -- ^ dischargeCekValEnv
+  -- frames
+  | FrameList -- ^ aka Backtrace
+  | FrameFocus Word -- ^ switch to frame by index no.
+  -- Q: Is exit really needed?
+  -- Wouldn't it be simpler for the Client to just ignore/throw away the debugger computation?
+  | Exit -- ^ exit button or sth else control-d
+
+type ShowUniFun uni fun =
+    ( uni `Everywhere` Show
+    , Closed uni
+    , GShow uni
+    , Show fun
+    )
+
+data DebuggerState uni fun = DebuggerState
+    { _dLoadedTerm :: DTerm uni fun -- it is only here so we can more easily Reload
+    , _dCekState :: CekState uni fun
+    , _dBreakpoints :: Seq Breakpoint
+    , _dFocusedFrameIx :: Word
+    }
+makeLenses ''DebuggerState
+
+-- | Given the debugger's state, have we reached a breakpoint?
+atBreakpoint :: DebuggerState uni fun -> Bool
+atBreakpoint newDState =
+    (newDState ^. dBreakpoints) `contain`
+    (newDState ^? dCekState . cekStateClosure . closureTerm . to termAnn) -- look at the SrcSpan of the Control term
+  where
+    -- FIXME: stub, implement this, could we use PlutusLedgerAPI.V1.Interval api for it?
+    contain :: Seq Breakpoint -> Maybe SrcSpans -> Bool
+    contain _bs = \case
+        Nothing -> False
+        Just _srcSpan -> False -- todo
+
+-- | The debugger's suspension functor
+data DebugF uni fun a
+  = InputF (Cmd -> a) -- iteratee
+  | LogF String a -- generator
+  | StepF
+      (CekState uni fun) -- ^ newState
+      (CekState uni fun -> a) -- ^ newCensoredState to continue
+  | UpdateTuiF (DebuggerState uni fun) a -- generator
```

Comment:
IMO this module shouldn't mention "TUI" or "FocusedFrame". We need to be able to integrate the same debug CEK machine with different UIs, not the other way around. In the future we may create a VSCode extension via Debug Adapter Protocol, or make a web-based UI, etc.